xmtp_mls/src/worker.rs

```diff
@@ -63,21 +71,37 @@ impl WorkerRunner {
         self.factories.push(Arc::new(factory))
     }
 
-    pub fn spawn(&self) {
-        for factory in &self.factories {
-            let metric = self.metrics.lock().get(&factory.kind()).cloned();
-            let (worker, metrics) = factory.create(metric);
+    pub fn spawn(self: &Arc<Self>) {
+        let mut handle_lock = self.handle.lock();
+        if let Some(handle) = handle_lock.take() {
+            handle.abort_handle().end();
+        }
+
+        let this = self.clone();
+        let handle = xmtp_common::spawn(None, async move {
+            let mut futs = FuturesUnordered::new();
+
+            for factory in &this.factories {
+                let metric = this.metrics.lock().get(&factory.kind()).cloned();
+                let (worker, metrics) = factory.create(metric);
+
+                if let Some(metrics) = metrics {
+                    this.metrics.lock().insert(factory.kind(), metrics);
+                }
 
-            if let Some(metrics) = metrics {
-                self.metrics.lock().insert(factory.kind(), metrics);
+                if let Some(metrics) = worker.metrics() {
+                    let mut m = this.metrics.lock();
+                    m.insert(worker.kind(), metrics);
+                }
```

Comment:
There's probably a better way to handle metrics than getting them passed back and forth and potentially overriding the metrics from the factory with the worker.